### PR TITLE
Caching blobs in shared memory for faster I/O

### PR DESCRIPTION
**Pull request by gohar94**
# Context
This PR is part of the work described in azure/azure-functions-host#7310

# All SDK Contribution checklist:

This checklist is used to make sure that common guidelines for a pull request are followed.
- [ ] **Please open PR in `Draft` mode if it is:**
	- Work in progress or not intended to be merged.
	- Encountering multiple pipeline failures and working on fixes.
- [ ] If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.
- [ ] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md).**
- [ ] **The pull request does not introduce [breaking changes](https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/breaking-change-rules.md).**

### [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md#general-guidelines)
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md#testing-guidelines)
- [ ] Pull request includes test coverage for the included changes.

### [SDK Generation Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md#sdk-generation-guidelines)
- [ ] The generate.cmd file for the SDK has been updated with the version of AutoRest, as well as the commitid of your swagger spec or link to the swagger spec, used to generate the code. (Track 2 only)
- [ ] The `*.csproj` and `AssemblyInfo.cs` files have been updated with the new version of the SDK. Please double check nuget.org current release version.

## Additional management plane SDK specific contribution checklist: 
Note: Only applies to `Microsoft.Azure.Management.[RP]` or `Azure.ResourceManager.[RP]`
 
- [ ] Include updated [management metadata](https://github.com/Azure/azure-sdk-for-net/tree/master/eng/mgmt/mgmtmetadata).
- [ ] Update AzureRP.props to add/remove version info to maintain up to date API versions.

### Management plane SDK Troubleshooting
- If this is very first SDK for a services and you are adding new service folders directly under /SDK, please add `new service` label and/or contact assigned reviewer.
- If the check fails at the `Verify Code Generation` step, please ensure:
	- Do not modify any code in generated folders.
	- Do not selectively include/remove generated files in the PR.
	- Do use `generate.ps1/cmd` to generate this PR instead of calling `autorest` directly.
	Please pay attention to the @microsoft.csharp version output after running `generate.ps1`. If it is lower than current released version (2.3.82), please run it again as it should pull down the latest version.
	
	**Note: We have recently updated the PSH module called by `generate.ps1` to emit additional data. This would help reduce/eliminate the Code Verification check error. Please run following command**:

	    `dotnet msbuild eng/mgmt.proj /t:Util /p:UtilityName=InstallPsModules`

### Old outstanding PR cleanup
 Please note:
	If PRs (including draft) has been out for more than 60 days and there are no responses from our query or followups, they will be closed to maintain a concise list for our reviewers.

### CODE REVIEW

File: sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Blobs/src/Bindings/NotifyingBlobStream.cs
Comment (goiri) [inline sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Blobs/src/Bindings/NotifyingBlobStream.cs:7]: Im guessing leftover

**Comment by kasobol-msft** [comment]
I suggest to convert this PR to draft while it's work in progress. Otherwise this will keep notifying people. https://github.blog/2019-02-14-introducing-draft-pull-requests/ .

I also suggest to review the design before implementing this.

**Comment by gohar94** [comment]
> I suggest to convert this PR to draft while it's work in progress. Otherwise this will keep notifying people. https://github.blog/2019-02-14-introducing-draft-pull-requests/ .
> 
> I also suggest to review the design before implementing this.

@kasobol-msft Thanks - changed back to draft. What in particular are you referring to in terms of design?

**Comment by kasobol-msft** [comment]
For design, I suggest to start internal email thread with me, @fabiocav and @paulbatum .

File: sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Blobs/src/Bindings/CacheAfterWriteStream.cs
Comment (kasobol-msft) [inline sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Blobs/src/Bindings/CacheAfterWriteStream.cs:78]: You can leverage https://github.com/Azure/azure-sdk-for-net/blob/17a6cbdeff99e7e881e27cc30da68e649f36966f/sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Blobs/src/DelegatingStream.cs#L11 to simplify this class. Similarly to https://github.com/Azure/azure-sdk-for-net/blob/17a6cbdeff99e7e881e27cc30da68e649f36966f/sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Blobs/src/Bindings/NotifyingBlobStream.cs#L10

File: sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Blobs/src/Bindings/CacheAfterWriteStream.cs
Comment (kasobol-msft) [inline sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Blobs/src/Bindings/CacheAfterWriteStream.cs:20]: The `Stream` and `ICacheAfterWriteStream` seem to be quite orthogonal concepts. Combining them like this make this type not very cohesive. One sign of this is that there's nothing in this class that would mix `Stream` and `ICacheAfterWriteStream`.

If the goal is to carry extra data along the stream then I'd suggest to use composition, i.e. create type enclosing Stream and these extra properties. This might involve some work on the binding configuration but would be less hacky.

File: sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Blobs/src/Bindings/CacheAfterWriteStream.cs
Comment (kasobol-msft) [inline sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Blobs/src/Bindings/CacheAfterWriteStream.cs:32]: Are we caching this because we want to call some API? or do we need some information about the blob.
I'm asking because this one might be quite heavy to cache (it has HttpPipeline and HttpClient inside).
Other concern would be that this has credentials inside - is it fine to cache? how long does cache live?
If this is long living and managed identity (oauth) is introduced then this thing will be refreshing token in background - is it fine?

File: sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Blobs/src/Config/BlobsExtensionConfigProvider.cs
Comment (kasobol-msft) [inline sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Blobs/src/Config/BlobsExtensionConfigProvider.cs:79]: do we need two constructors? 
if yes then one should call into another.

File: sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Blobs/src/Config/BlobsExtensionConfigProvider.cs
Comment (kasobol-msft) [inline sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Blobs/src/Config/BlobsExtensionConfigProvider.cs:368]: should we check some runtime value instead of this to detect where this code is running?
I'd rather not use absence of that cache to tell where we are. Is IFunctionDataCache a public type? - if yes then user of the extension can trigger this logic by providing own imlementation.

File: sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Blobs/src/ReadBlobArgumentBinding.cs
Comment (kasobol-msft) [inline sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Blobs/src/ReadBlobArgumentBinding.cs:83]: keep in mind that this call and OpenRead might see different state of the blob. These are two separate calls to storage backend, so there might be a race condition here.
It might be better to fetch properties first and openRead with retrieved Etag later, so that both are assured to see same view of the blob (or fail).

**Review by kasobol-msft — COMMENTED** [review]
Thank you for the context provided offline. This helped a lot to understand this PR.

Could you please update description of this PR to reference related work? 
Could you also provide a short write up explaining the change for other reviewers?

File: sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Blobs/src/Bindings/CacheObjectAndBlobStream.cs
Comment (kasobol-msft) [inline sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Blobs/src/Bindings/CacheObjectAndBlobStream.cs:16]: nit: I'd call this `CacheableWriteBlob` and the other one `CacheableReadBlob`

File: sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Blobs/src/ReadBlobArgumentBinding.cs
Comment (kasobol-msft) [inline sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Blobs/src/ReadBlobArgumentBinding.cs:45]: This overload looks almost like the other one. I'd rather add optional `eTag` to other overload and create conditions if it's provided.

**Review by kasobol-msft — COMMENTED** [review]
This looks good. Please add test coverage and clean up changes in csproj and .Config files.

File: sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Blobs/src/ReadBlobArgumentBinding.cs
Comment (goiri) [inline sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Blobs/src/ReadBlobArgumentBinding.cs:64]: Mention in the comment why we set to false the configure await.

File: sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Blobs/tests/Bindings/CacheableReadBlobTests.cs
Comment (goiri) [inline sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Blobs/tests/Bindings/CacheableReadBlobTests.cs:21]: Add a comment explaining why this implies a hit and  39 does not.

File: sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Blobs/tests/Bindings/CacheableReadBlobTests.cs
Comment (goiri) [inline sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Blobs/tests/Bindings/CacheableReadBlobTests.cs:96]: Assert.False

**Comment by gohar94** [comment]
@kasobol-msft Ready for your final pass. This PR will still have to wait for the WebJobs SDK changes to go in though since that's where the new interfaces are defined.

File: sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Blobs/tests/Bindings/CacheableReadBlobTests.cs
Comment (goiri) [inline sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Blobs/tests/Bindings/CacheableReadBlobTests.cs:116]: You use it before but here might be a good place to explain what the cache mock verifies as an assert

File: sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Blobs/tests/Bindings/CacheableWriteBlobTests.cs
Comment (goiri) [inline sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Blobs/tests/Bindings/CacheableWriteBlobTests.cs:102]: False

File: sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Blobs/tests/Bindings/CacheableWriteBlobTests.cs
Comment (goiri) [inline sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Blobs/tests/Bindings/CacheableWriteBlobTests.cs:118]: A comment with the high level purpose for each static method.

File: sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Queues/src/Microsoft.Azure.WebJobs.Extensions.Storage.Queues.csproj
Comment (goiri) [inline sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Queues/src/Microsoft.Azure.WebJobs.Extensions.Storage.Queues.csproj]: You may want to checkout from origin to get rid of this noop change.

File: sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Blobs/src/ReadBlobArgumentBinding.cs
Comment (kasobol-msft) [reply sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Blobs/src/ReadBlobArgumentBinding.cs:64]: I'd not do this. There's analyzer plugged into build that makes sure every `awat Foo().ConfigureAwait(false)` does this.

File: sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Queues/src/Microsoft.Azure.WebJobs.Extensions.Storage.Queues.csproj
Comment (gohar94) [reply sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Queues/src/Microsoft.Azure.WebJobs.Extensions.Storage.Queues.csproj]: Yes this is gone now

**Comment by kasobol-msft** [comment]
> @kasobol-msft Kamil Sobol FTE Ready for your final pass. This PR will still have to wait for the WebJobs SDK changes to go in though since that's where the new interfaces are defined.

Looks good. Let's have one more when dependencies are resolved and build/ci is passing.

Could you also add some E2E tests (like here https://github.com/Azure/azure-sdk-for-net/tree/master/sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Scenario.Tests/tests) that cover sunny day scenario with cache enabled (cache can be a Mock for this purpose, but I'd like to see it working e2e). Existing E2E scenarios should cover case where cache isn't enabled (if we leave them alone).

**Comment by gohar94** [comment]
@kasobol-msft We are closing in on the other PRs for this work so it would be good to get a final review/approval from you on this now. Thanks a lot.

**Review by goiri — APPROVED** [review]

File: sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Scenario.Tests/tests/CacheableBlobsEndToEndTests.cs
Comment (kasobol-msft) [inline sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Scenario.Tests/tests/CacheableBlobsEndToEndTests.cs:121]: what's the assertion here? lack of exception thrown?

**Review by kasobol-msft — COMMENTED** [review]
LGTM, please reach out again when checks are all green.

File: sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Scenario.Tests/tests/CacheableBlobsEndToEndTests.cs
Comment (gohar94) [reply sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Scenario.Tests/tests/CacheableBlobsEndToEndTests.cs:121]: The assertion is inside the function `ByteArrayBindingWriteToCacheAsync` that will be called. So if that fails, then this throws.

**Comment by gohar94** [comment]
> LGTM, please reach out again when checks are all green.

@kasobol-msft I have added the myget.org feed to `nuget.conf` - I see there is a comment restricting the addition of further nuget feeds but this is required to pick up the latest webjobs package we have published to myget which is not published on nuget.org yet.
For further context, we can discuss with Naren Soni on the Functions team as he is the one who provided me this information.

**Comment by gohar94** [comment]
> Hi @gohar94. Thank you, for your interest in helping to improve the Azure SDK experience and for your contribution. We've noticed that there hasn't been recent engagement on this pull request. If this is still an active work stream, please let us know by pushing some changes or leaving a comment. Otherwise, we'll close this out in 7 days.

...

**Review by kasobol-msft — APPROVED** [review]

**Pull request merged**
